impr: S3UTILS-139 repairDuplicateVersionIds

jonathan-gramain opened the pull request:
New script derived from repairDuplicateVersions to repair object metadata that has a duplicate versionId field, notably due to bug S3C-7861 (CRR with incompatible S3C versions between source and target).

It takes log lines from verifyBucketSproxydKeys as input, and for each line matching a duplicate versionId detection, repairs the affected object.

jonathan-gramain committed:
impr: S3UTILS-139 repairDuplicateVersionIds

New script derived from repairDuplicateVersions to repair object
metadata that has a duplicate versionId field, notably due to bug
S3C-7861 (CRR with incompatible S3C versions between source and
target).

It takes log lines from verifyBucketSproxydKeys as input, and for each
line matching a duplicate versionId detection, repairs the affected
object.

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/S3UTILS-139-repairDuplicateVersionIds` into
`development/1.13`
* [w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds](https://github.com/scality/s3utils/compare/development/1.14...w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds) will be merged into `development/1.14` 



The following branches will **NOT** be impacted:

* `development/1.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #38a973ce59689e59e0316827334924b4efa14912 on the integration branch
`w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds` is merging a branch which is neither the current
branch `improvement/S3UTILS-139-repairDuplicateVersionIds` nor the development branch
`development/1.14`.

It is likely due to a rebase of the branch `improvement/S3UTILS-139-repairDuplicateVersionIds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/S3UTILS-139-repairDuplicateVersionIds` into
`development/1.13`
* [w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds](https://github.com/scality/s3utils/compare/development/1.14...w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds) will be merged into `development/1.14` 



The following branches will **NOT** be impacted:

* `development/1.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

KazToozs reviewed (approved):
cool signal handling for the exiting, not something I'd seen yet in JS. LGTM otherwise

jonathan-gramain committed:
S3UTILS-139 repairDuplicateVersionIds: safer repair

Add safety to repair operation:

- check the state of the master key before repairing, to ensure there
  is still a duplicate "versionId" field

- repair using the "versionId" just read from the metadata, rather
  than the one from the log, to make sure we repair an up-to-date
  version (or be reasonably sure about it)

bert-e commented:
# History mismatch



Merge commit #f22ab5db422b20d6d65549139c85261b4f32ebe4 on the integration branch
`w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds` is merging a branch which is neither the current
branch `improvement/S3UTILS-139-repairDuplicateVersionIds` nor the development branch
`development/1.14`.

It is likely due to a rebase of the branch `improvement/S3UTILS-139-repairDuplicateVersionIds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
@KazToozs @tmacro I added an extra commit to improve the safety of the repair, indeed, previously it would be able to repair a master with the original scanned version ID, even if the master key was updated in the meantime.

jonathan-gramain commented:
/approve

bert-e commented:
# History mismatch



Merge commit #f22ab5db422b20d6d65549139c85261b4f32ebe4 on the integration branch
`w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds` is merging a branch which is neither the current
branch `improvement/S3UTILS-139-repairDuplicateVersionIds` nor the development branch
`development/1.14`.

It is likely due to a rebase of the branch `improvement/S3UTILS-139-repairDuplicateVersionIds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

jonathan-gramain commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/S3UTILS-139-repairDuplicateVersionIds` into
`development/1.13`
* [w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds](https://github.com/scality/s3utils/compare/development/1.14...w/1.14/improvement/S3UTILS-139-repairDuplicateVersionIds) will be merged into `development/1.14` 



The following branches will **NOT** be impacted:

* `development/1.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`

* :heavy_check_mark: `development/1.14`



The following branches will **NOT** be impacted:

* `development/1.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`

* :heavy_check_mark: `development/1.14`



The following branches have **NOT** changed:

* `development/1.4`



Please check the status of the associated issue S3UTILS-139.

Goodbye jonathan-gramain.
